yt/utilities/lib/interpolators.pyx

```diff
@@ -163,3 +163,115 @@ def ghost_zone_interpolate(int rf,
                 opos[2] += ods[2]
             opos[1] += ods[1]
         opos[0] += ods[0]
+
+@cython.cdivision(True)
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def replace_nonperiodic_with_extrap(np.ndarray[np.float64_t, ndim=3] field,
+                                    np.ndarray[np.int_t,     ndim=1] ncell_lo,
+                                    np.ndarray[np.int_t,     ndim=1] ncell_hi):
+    """
+    The purpose of this function is to extrapolate data from the interior of
+    the buffer region to replace the given data in the buffer region using a
+    first-order approximation.
+
+    Parameters
+    ----------
+    field : 3D array
+        Three-dimensional data of the flowfield
+    ncell_lo : 1D array of length 3
+        Number of buffer cells to replace on the low side along each axis
+    ncell_hi : 1D array of length 3
+        Number of buffer cells to replace on the high side along each axis
+
+    This is best illustrated through a one-dimensional example. Consider a
+    linear function: f(x) = x , and the `field` variable was filled using
+    periodicity, so we have the following data:
+
+    -----------------------------------------------
+    index:        0  |  1  |  2  |  3  |  4  |  5
+    -----------------------------------------------
+    x or f(x):  -0.5 | 0.5 | 1.5 | 2.5 | 3.5 | 4.5
+    -----------------------------------------------
+    field:       3.5 | 0.5 | 1.5 | 2.5 | 3.5 | 0.5
+    -----------------------------------------------
+
+    where field.shape ---> (6,1,1), ncell_lo ---> (1,0,0), and
+    ncell_hi ---> (1,0,0) . For this example, we replace the value of
+    field[0,0,0] using linear extrapolation from field[1,0,0] and field[2,0,0],
+    similarly for upper boundary. Evaluating this function then yields the
+    following output of `field`:
+
+    -------------------------------------------
+    field:   -0.5 | 0.5 | 1.5 | 2.5 | 3.5 | 4.5
+    -------------------------------------------
+
+    This is particularly important for estimating gradients near non-periodic
+    boundaries in the domain.
+
+    """
+    cdef int i, j, k
+    cdef int ii, ij, ik
+    cdef int nx, ny, nz
+    cdef int lx, ly, lz
+    cdef int hx, hy, hz
+    cdef np.float64_t xp, xm, yp, ym, zp, zm
+    cdef np.float64_t qp, qm
+
+    nx, ny, nz = (field.shape[0], field.shape[1], field.shape[2])
```

Comment:
assuming this is valid
```suggestion
    nx, ny, nz = field.shape
```

Reply:
Unfortunately this is not valid in cython.